config/configqueue/queue.go

```diff
@@ -0,0 +1,69 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package configqueue // import "go.opentelemetry.io/collector/config/configqueue"
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+// QueueConfig defines configuration for queueing requests before exporting.
+// It's supposed to be used with the new exporter helpers New[Traces|Metrics|Logs]RequestExporter.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type QueueConfig struct {
+	// Enabled indicates whether to not enqueue batches before exporting.
+	Enabled bool `mapstructure:"enabled"`
+	// NumConsumers is the number of consumers from the queue.
+	NumConsumers int `mapstructure:"num_consumers"`
+	// QueueSizeItems is the maximum number of items (spans, metric data points or log records)
+	// allowed in queue at any given time.
+	QueueSizeItems int `mapstructure:"queue_size_items"`
+	// QueueSize is the maximum number of requests allowed in queue at any given time.
+	// This option is left for backward compatibility and will be deprecated in the future.
+	// It's recommended to use QueueSizeItems instead.
+	QueueSize int `mapstructure:"queue_size"`
+}
+
+// NewDefaultQueueConfig returns the default Config.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+func NewDefaultQueueConfig() QueueConfig {
+	return QueueConfig{
+		Enabled:        true,
+		NumConsumers:   10,
+		QueueSizeItems: 100_000,
+	}
+}
+
+// Validate checks if the QueueSettings configuration is valid
+func (qCfg *QueueConfig) Validate() error {
+	if !qCfg.Enabled {
+		return nil
+	}
+	if qCfg.NumConsumers <= 0 {
+		return errors.New("number of consumers must be positive")
+	}
+	if qCfg.QueueSizeItems > 0 && qCfg.QueueSize > 0 {
+		return errors.New("only one of 'queue_size' and 'queue_size_items' can be specified")
+	}
+	if qCfg.QueueSizeItems <= 0 && qCfg.QueueSize <= 0 {
+		return errors.New("queue size must be positive")
+	}
+	return nil
+}
+
+// PersistentQueueConfig defines configuration for queueing requests in a persistent storage.
+// The struct is provided to be added in the exporter configuration as one struct under the "sending_queue" key.
+// The exporter helper Go interface requires the fields to be provided separately to WithRequestQueue and
+// NewPersistentQueueFactory.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type PersistentQueueConfig struct {
```

Comment:
Would be interested to see how people decide to use this vs QueueConfig in their Configs. When do I pick one or the other?

Reply:
`QueueConfig` is used only for exporters that don't support persistent queue. And `PersistentQueueConfig` for exporters that support both. Maybe we don't need to support memory-queue-only exporters. My idea was that some exporters don't want to provide serializers but still want to use custom requests.

Reply:
>  My idea was that some exporters don't want to provide serializers but still want to use custom requests.

We can have some default serializers using pdata maybe in that case? May complicate too much the implementation design...

Reply:
That would require asking user for Request->pdata translator. Right now we ask for pdata->Request only

Reply:
Otherwise, we can apply the conversion after the queue. This can significantly complicate the implementation, as you mentioned.